gdnative-core/src/geom/basis.rs

```diff
@@ -29,14 +116,316 @@ impl Basis {
+    pub fn determinant(&self) -> f32 {
+        let elements = self.elements;
+        return elements[0].x * (elements[1].y * elements[2].z - elements[2].y * elements[1].z)
+            - elements[1].x * (elements[0].y * elements[2].z - elements[2].y * elements[0].z)
+            + elements[2].x * (elements[0].y * elements[1].z - elements[1].y * elements[0].z);
+    }
+
+    /// Orthonormalizes the matrix
+    ///
+    /// Performs a [Gram-Schmidt orthonormalization](https://en.wikipedia.org/wiki/Gram-Schmidt_process) on the basis of the matrix.
+    /// This can be useful to call from time to time to avoid rounding error for orthogonal matrices.
+    pub fn orthonormalize(&mut self) {
+        assert!(
+            !self.determinant().approx_eq(&0.0),
+            "Determinant should not be zero."
+        );
+
+        // Gram-Schmidt Process
+        let mut x = self.x();
+        let mut y = self.y();
+        let mut z = self.z();
+
+        x = x.normalize();
+        y = y - x * (x.dot(y));
+        y = y.normalize();
+        z = z - x * (x.dot(z)) - y * (y.dot(z));
+        z = z.normalize();
+
+        self.set_x(x);
+        self.set_y(y);
+        self.set_z(z);
+    }
+
+    /// Returns an orthonormalized version of the matrix.
+    ///
+    /// See [`Basis::orthonormalize()`](./struct.Basis.html#method.orthonormalize)
+    pub fn orthonormalized(mut self) -> Basis {
+        self.orthonormalize();
+        self
+    }
+
+    /// Returns true if this basis and b are approximately equal.
+    pub fn approx_eq(&self, other: &Basis) -> bool {
+        return self.elements[0].approx_eq(&other.elements[0])
+            && self.elements[1].approx_eq(&other.elements[1])
+            && self.elements[2].approx_eq(&other.elements[2]);
+    }
+
+    fn is_orthogonal(&self) -> bool {
+        let identity = Self::identity();
+        let m = (*self) * self.transposed();
+        m.approx_eq(&identity)
+    }
+
+    fn is_rotation(&self) -> bool {
+        let det = self.determinant();
+        det.approx_eq(&1.0) && self.is_orthogonal()
+    }
+
+    /// Multiplies the matrix from left by the rotation matrix: M -> R.M
+    ///
+    /// The main use of Basis is as Transform.basis, which is used a the transformation matrix
+    /// of 3D object. Rotate here refers to rotation of the object (which is R * (*this)),
+    pub fn rotated(self, axis: &Vector3, phi: f32) -> Basis {
+        let rot = Basis::from_axis_angle(&axis, phi);
+        rot * self
+    }
+
+    /// Rotates the matrix.
+    ///
+    /// See [`Basis::rotated()`](./struct.Basis.html#method.rotated)
+    pub fn rotate(&mut self, axis: &Vector3, phi: f32) {
+        *self = self.rotated(axis, phi);
+    }
+
+    /// The basis’s rotation in the form of a quaternion.
+    ///
+    /// See [`Basis::euler()`](./struct.Basis.html#method.euler) if you need Euler angles, but keep in mind quaternions should generally be preferred to Euler angles.
```

Comment:
As said in an earlier comment, we don't teach people how to code. Instead, state the facts and let the users decide for themselves. This should be reworded as:

> Quaternions are frequently used in 3D graphics, because they enable easy and cheap interpolation. However, they are less human-readable. For Euler angles, see [`Basis::euler`](./struct.Basis.html#method.euler).